pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -862,7 +863,9 @@ private Map<Integer, SegmentZKMetadata> getLatestSegmentZKMetadataMap(String rea
    * TODO: We need to find a place to detect and update a gauge for nonConsumingPartitionsCount for a table, and
    * reset it to 0 at the end of validateLLC
    */
-  public void ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevelStreamConfig streamConfig) {
+  // TODO(saurabh) : We can reduce critical section by having tableName level locks?
+  public synchronized void ensureAllPartitionsConsuming(TableConfig tableConfig,
```

Comment:
Let's update the javadoc for this method to include the new scenario:
```
If the consuming segment is deleted:
Check whether there are segments in the PROPERTYSTORE with status DONE, but no new segment in status IN_PROGRESS, and the state for the latest segment in the IDEALSTATE is ONLINE
```
(Note that this is very similar to the failure between step-1 and step-2, the only difference is the state in the ideal state)

Reply:
> It should work without synchronization because it is performing a check and write

Wasn't aware of this. Looks like helix itself guarantees concurrent updates are handeled safely. Removed synchronization. 



> Let's update the javadoc for this method to include the new scenario:
> 
> ```
> If the consuming segment is deleted:
> Check whether there are segments in the PROPERTYSTORE with status DONE, but no new segment in status IN_PROGRESS, and the state for the latest segment in the IDEALSTATE is ONLINE
> ```
> 
> (Note that this is very similar to the failure between step-1 and step-2, the only difference is the state in the ideal state)

Ack. Added.